src/bin/wasmer.rs

```diff
@@ -129,48 +194,81 @@ fn execute_wasm(options: &Run) -> Result<(), String> {
-            wasmer_emscripten::generate_emscripten_env(&mut emscripten_globals),
-            Some(emscripten_globals), // TODO Em Globals is here to extend, lifetime, find better solution
-        )
-    } else {
-        (
-            InstanceABI::None,
-            wasmer_runtime_core::import::ImportObject::new(),
-            None,
-        )
+    let key = String::from(options.path.to_str().unwrap());
+    run_wasm_module(&module, key, &options.args)
+}
+
+fn run(options: Run) {
+    match execute_wasm(&options) {
+        Ok(()) => {}
+        Err(message) => {
+            eprintln!("{:?}", message);
+            exit(1);
+        }
+    }
+}
+
+fn generate_wasm_cache(options: &CacheGenerate) -> Result<(), String> {
+    let wasm_binary = load_wasm_binary(&options.path)?;
+    let module = webassembly::compile(&wasm_binary[..])
+        .map_err(|e| format!("Can't compile module: {:?}", e))?;
+
+    // We generate a hash for the given binary, so we can use it as key
+    // for the Filesystem cache
+    let hash = WasmHash::generate(&wasm_binary);
+
+    let wasmer_cache_dir = get_cache_dir();
+
+    // We create a new cache instance.
+    // It could be possible to use any other kinds of caching, as long as they
+    // implement the Cache trait (with save and load functions)
+    let mut cache = unsafe {
+        FileSystemCache::new(wasmer_cache_dir).map_err(|e| format!("Cache error: {:?}", e))?
     };
 
-    let mut instance = module
-        .instantiate(&import_object)
-        .map_err(|e| format!("Can't instantiate module: {:?}", e))?;
+    // We save the module into a cache file
+    cache
+        .store(hash.encode(), module.clone())
+        .map_err(|e| format!("Can't store module in cache: {:?}", e))
+}
 
-    webassembly::run_instance(
-        &module,
-        &mut instance,
-        options.path.to_str().unwrap(),
-        options.args.iter().map(|arg| arg.as_str()).collect(),
-    )
-    .map_err(|e| format!("{:?}", e))?;
+fn cache_generate(options: CacheGenerate) {
+    match generate_wasm_cache(&options) {
+        Ok(()) => {}
+        Err(message) => {
+            eprintln!("{:?}", message);
+            exit(1);
+        }
+    }
+}
 
-    Ok(())
+fn run_wasm_module_from_cache(options: &CacheRun) -> Result<(), String> {
+    let wasmer_cache_dir = get_cache_dir();
+
+    // We create a new cache instance.
+    // It could be possible to use any other kinds of caching, as long as they
+    // implement the Cache trait (with save and load functions)
+    let cache = unsafe {
+        FileSystemCache::new(wasmer_cache_dir).map_err(|e| format!("Cache error: {:?}", e))?
+    };
+
+    let key = String::from(options.module_key.to_str().unwrap());
+    let module = cache
+        .load(key)
+        .map_err(|e| format!("Can't execute module from cache: {:?}", e))?;
+
+    let name = String::from(options.module_name.to_str().unwrap());
+    run_wasm_module(&module, name, &options.args)
```

Comment:
A small nitpick: the `module_name` is unecessary on the CLI. Can you pass the hash as a string to `run_wasm_module` in place of the name? This will allow us to remove the `module_name` field from `CacheRun` too. 

This way all we have to type is:
`$ wasmer cache run VERY_LONG_HASH -- -c /opt/nginx/config`

Reply:
I can do that, I was not sure if it'd be ok :+1: